### PR TITLE
DNS resolver issues

### PR DESCRIPTION
**Pull request by vietj**
- resolve localhost to 127.0.0.1 on Windows as Windows 10
- fix bug with resolver with a DNS answer containing CNAME+A

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -420,6 +420,18 @@
               </systemProperties>
             </configuration>
           </execution>
+          <execution>
+            <id>window-resolve-localhost</id>
```
Comment (tsegismont) [inline pom.xml]: window**s** ?

File: src/test/java/io/vertx/test/core/HostnameResolutionTest.java [hunk 2]
```diff
@@ -309,4 +310,22 @@ public void testResolveFromBuffer() {
     }));
     await();
   }
+
+/*
```
Comment (tsegismont) [inline src/test/java/io/vertx/test/core/HostnameResolutionTest.java]: Remove this ?

**Comment by alexlehm** [comment]
this is probably dependent on the hosts file contains an entry for localhost 127.0.0.1 since the hosts file is evaluated by Netty resolver.
I didn't notice that before, since my windows at home apparently contains an entry for that (windows7), but the machine I am using at work doesn't (also windows7)

**Comment by vietj** [comment]
FTR:

https://github.com/netty/netty/issues/5386
https://github.com/netty/netty/issues/5391

File: src/main/java/io/vertx/core/impl/AddressResolver.java [hunk 3]
```diff
@@ -131,14 +133,18 @@ public void close() {
...
-    io.netty.util.concurrent.Future<InetSocketAddress> fut = resolver.resolve(InetSocketAddress.createUnresolved(hostname, 80));
+    io.netty.util.concurrent.Future<InetSocketAddress> fut = resolver.resolve(InetSocketAddress.createUnresolved(hostname, 0));
     fut.addListener(a -> {
       callback.runOnContext(v -> {
         if (a.isSuccess()) {
           InetSocketAddress address = fut.getNow();
           resultHandler.handle(Future.succeededFuture(address.getAddress()));
         } else {
-          resultHandler.handle(Future.failedFuture(a.cause()));
+          if (Utils.isWindows() && hostname.equals("localhost")) {
```
Comment (alexlehm) [inline src/main/java/io/vertx/core/impl/AddressResolver.java]: hostname resolution is case-insensitive so LOCALHOST is a valid hostname for 127.0.0.1 (had the same issue in netty hosts file resolver)

hostname.toLowerCase(Locale.EN).equals("localhost");

Comment (vietj) [inline src/main/java/io/vertx/core/impl/AddressResolver.java]: in the more recent version there is not comparison anymore, what is done instead is to add "alias=127.0.0.1" in the "hosts" Map, so I think this is handled by the underlying resolver, can you check ?

**Comment by alexlehm** [comment]
this works correctly, the entry is added when localhost is not listed in the hosts file and the comparison in correct case-insensitive in the netty code.

**Comment by alexlehm** [comment]
For some reason I cannot reproduce the failing resolution with LOCALHOST anymore when I take out the code to add the localhost alias

**Pull request merged**